### PR TITLE
extend genTuple to wrap the rhs of an infixapp in parens if needed

### LINKED ISSUE #2819: If then expression inside object instantiation breaks when formatted
Issue created from [fantomas-online](https://fsprojects.github.io/fantomas-tools/#/fantomas/preview?data=N4KABGBEDGD2AmBTSAuKAXAngB0WAwgBYCG26iATgBQCUYAvADoB2EEAtouwEaVgBuxADZgAKgEt0QxGgDO6CuOYBzMLDLjYremACqzaIUTQA1ongA6JADNiAVyHpY1gDwB9AHxgA7pMJhlRHQAGjBZIJY2ME4ePkERXQohOQUlVR19Q2MzSxt7R2d3L190f0CQsIjmFml0MGgGSLYiUnJqJqiICSk8JlZOgfFrMAU7PFLEav6BmbAAZVhOMEZIFY7ZsEQhcPWNtgA5LURdmeCTiESRHTXV6bAaFkhg8Chw9HQ02VQwAG0XiFAA0gABIsLhvpAlOgnv82CD4mNvn9ZgAGZ6zSAASWYSGY6Dm4gAXsh0TMACywiAAXVhAF9SQDKVBQThkGhIXiYUDgQi2b8mRAAIwMzqQACyxAAHgAZJSIaWTZSlLkzQUAJhRTJpUXpsMBopZ4PZk3gAHlrLLmIg5lhpCqovDhIi0MiZmqRQ6AKI482WkkCqBCayQLV0kX6h2GvmQbiwWBCe1wnlOvmugYAZg9cOx4Qo6AAYkphPtEN4hHLE1FRohQzrw0yQWDo7H45XmbykQGyVmIJA5thiNBEAAhRDWWAURAABWIFGInDabYg1drbF1UQjSabEJbCazjqEzv5swArD2oP3ByOxxP5bBvJRoMRwtj%2BLAnx8tEuRhQxquIOubCbr2UY7nGe4Nsmh6pgGABs559gOQ6juOk66NguAUE%2BL7MG%2BH6aMw362NsNYDNqa71ty27srubYHkeaadAA7Ahl7ITek74EIz6yPgWjyL%2B0BOBQRHCOE-5gIBjJUayYGtvuUEMQGAAcrFIdeqGIGKXC8CJ57EeJZFhnqkHUVAtEKR2LoBgAnGpV4obefFCF%2B%2BliaRnTkQBlEGmZMbgXRikwbMgpogGiFXgAgtYbR8ew7DEN%2BK5GXWJkyUa5kBZZKadiFwrhWxGm3nMXDiHALmEW5JESVJYDAcyfkWZBVnHqq7oFep0VtCV7BlfGrkBslnnGRupmyTRWXNTl1khZmHVRRQsB2DiAAiWziL1i7nkNUReZJPmRn5UKBS1jFRIK3bhRKkqYtYohGMwcyEBO6AAOriPAyrnpqKUUWlvnjVAx3ZdBuWqmeV1Srd92TJ6JFPS972fYQ36wT9w2paN6XRsDU2gzNqrwZDN3MNY4iSqamHEMJnqStgk6yLIBHfsp6O7SNQFjRlHLQiDSkhSxxMAErGBO8BI19XZs2we21fVjaAzzJ3Ta1AyCqpwuixQ8D7HYsQUOamLkOwXznoKNUHVuivsA4Hzlla%2BYTgl7yUKIsl88Fqp2eFItwNrYq2%2BI9uII7FDO1t4WGLOg5tG4vjIyGv3ef9h2K7j3KnQGGoIddkUUHOmCmhQsryBLKPnqzFsp1b3PpwamezGq%2BUYrn%2BfEIXxfiPIuv64bxumwG5tJ-t1cgX5NuOEHcqh%2BHrvu3j-Nuu1GJ5wXRcl%2BgAeT8HM-UxHGJR3OQmUHHH3KlXWMA7XnIe2DAxqnNLdSgAaimw5KPAaRlyz0vUhz0lXxxjfBent76XSfpKfMy0hIEXfjiL%2BZ9y5SwvpzbGEI66Rgbm6CGECVqwHQAAcSCLTemiBGYEW-hXX%2BYBZaWzHorJqGcVZnTYGqImGISxljlKOdAD5Jhu1wGtMmzBJAEUijibS%2BsB6zB2jLf%2BdUubNkmkw-GqtOhqkFivcsyhmBQIMJ%2BR64gdHUzsJOUQsBsS4nQNTZmVVDIYz%2BpfVO3NGH12YVnDWWi2jMBsfwO8KhJG6SEUoUR-FRLVWHnLRRckIIqMXvfb2LdA5%2BmlPObg8BiBcVgEzFQnDg7hPsezTGqDAExOVqolhEB0xhQxCQygG1JjWKEAAaUQIgbAljGnYmHHOAwSDZgGQ8kUxxJTnFKPkiAu%2BnR0zNxmJAYc3FmAmD9LIPOS0cQlnkOYLeds5QkIZkzMJ21fxDLkcUgBYyym3wJhmZeczhyzkcpOFaXdoCKF6j48g8B9AETWtAbic4DEFNOX-c5Ci0Hsgnrsq0PTBxmHxLaf0cTQHTMfnMnZU8YVH3hTaTAdoEJvPnLgeAicHHJycTXIBvNJk3OmeAuZrT2nXR7rpc0CziBLJWd%2BUK1DaGjwagw5RbiKkBnTDguZeTuEcS0oHYOcVsB2GsQY-ZZDDmVUGiclBFzKVXJpWoqI6Z2FzLqe8xpwg4a9VkP0mYgyaovCpKSSEsh8xM2%2BLa2kQA)

#### Code

```fsharp
type Chapter() =
    member val Title: string option = Unchecked.defaultof<_> with get, set
    member val Url: string = Unchecked.defaultof<_> with get, set

let c =
    Chapter(
        Title =
            if true then
                Some ""
            else
                None
            ,
        Url = ""
    )

```
            

#### Result

```fsharp
type Chapter() =
    member val Title: string option = Unchecked.defaultof<_> with get, set
    member val Url: string = Unchecked.defaultof<_> with get, set

let c = Chapter(Title = if true then Some "" else None, Url = "")

```
            
#### Problem description

After formatting, this code no longer compiles. I think the issue is ambiguity with creating a tuple when the comma is on the same line as the if expression. Inserting a newline or wrapping the if expression in parenthesis fixes the compilation error.

#### Extra information

- [x] The formatted result breaks my code.
- [x] The formatted result gives compiler warnings.
- [ ] I or my company would be willing to help fix this.

#### Options

Fantomas v5.2 branch at 1/1/1990

Default Fantomas configuration

<sub>Did you know that you can ignore files when formatting from fantomas-tool or the FAKE targets by using a [.fantomasignore file](https://fsprojects.github.io/fantomas/docs/end-users/IgnoreFiles.html)?</sub>

### PR DESCRIPTION
**Pull request by dawedawe**
Fixes #2819

### CODE REVIEW

File: src/Fantomas.Core/CodePrinter.fs [hunk 1]
```diff
@@ -300,14 +300,19 @@ let genInheritConstructor (ic: InheritConstructor) =
         genType node.Type
         +> sepSpaceOrIndentAndNlnIfExpressionExceedsPageWidth (genExpr node.Expr)
 
-let isSynExprLambdaOrIfThenElse (e: Expr) =
+let isSynExprIfThenElse (e: Expr) =
```
Comment (nojaf) [inline src/Fantomas.Core/CodePrinter.fs:303]: I think we want these functions in two flavours: partial active pattern and predicate function.
I would also not use `SynExpr` in the name here.

File: src/Fantomas.Core/CodePrinter.fs [hunk 2]
```diff
@@ -1834,15 +1828,36 @@ let genMultilineFunctionApplicationArguments (argExpr: Expr) =
     | _ -> genExpr argExpr
 
 let genTuple (node: ExprTupleNode) =
+    let wrapInfixAppRhsInParenIfNeeded expr =
+        match expr with
+        | Expr.InfixApp exprInfixAppNode ->
+            match exprInfixAppNode.RightHandSide with
+            | e when isSynExprLambdaOrIfThenElse e ->
```
Comment (nojaf) [inline src/Fantomas.Core/CodePrinter.fs:1835]: Like here, the partial active pattern would be nicer.

File: src/Fantomas.Core/CodePrinter.fs [hunk 3]
```diff
@@ -300,14 +300,19 @@ let genInheritConstructor (ic: InheritConstructor) =
...
-    | Expr.Lambda _
     | Expr.IfThen _
     | Expr.IfThenElif _
     | Expr.IfThenElse _ -> true
     | _ -> false
 
+let isSynExprLambdaOrIfThenElse (e: Expr) =
+    match e with
+    | Expr.Lambda _ -> true
+    | _ when isSynExprIfThenElse e -> true
```
Comment (nojaf) [inline src/Fantomas.Core/CodePrinter.fs:313]: I really dislike the wildcard pattern with a when clause.
This should be its own partial active pattern.

File: src/Fantomas.Core/CodePrinter.fs [hunk 4]
```diff
@@ -300,14 +300,19 @@ let genInheritConstructor (ic: InheritConstructor) =
...
+    | _ -> false
```
Comment (nojaf) [inline src/Fantomas.Core/CodePrinter.fs:314]: Or at the very least return `| _ -> isSynExprIfThenElse e`

File: src/Fantomas.Core/CodePrinter.fs [hunk 5]
```diff
@@ -1834,15 +1828,36 @@ let genMultilineFunctionApplicationArguments (argExpr: Expr) =
     | _ -> genExpr argExpr
 
 let genTuple (node: ExprTupleNode) =
+    let wrapInfixAppRhsInParenIfNeeded expr =
+        match expr with
+        | Expr.InfixApp exprInfixAppNode ->
+            match exprInfixAppNode.RightHandSide with
+            | e when isSynExprLambdaOrIfThenElse e ->
+                let parenNode =
+                    ExprParenNode(
```
Comment (nojaf) [inline src/Fantomas.Core/CodePrinter.fs:1837]: Check the usage for ExprPareNode, we already use this in CodePrinter, time to extract a helper for this:
image

File: src/Fantomas.Core/CodePrinter.fs [hunk 6]
```diff
@@ -1834,15 +1828,36 @@ let genMultilineFunctionApplicationArguments (argExpr: Expr) =
...
     let shortExpression =
         col sepNone node.Items (function
             | Choice1Of2 e ->
                 match e with
-                | Expr.IfThen _
-                | Expr.IfThenElif _
-                | Expr.IfThenElse _
-                | Expr.Lambda _ -> sepOpenT +> genExpr e +> sepCloseT
-                | e -> genExpr e
+                | e when isSynExprLambdaOrIfThenElse e -> sepOpenT +> genExpr e +> sepCloseT
```
Comment (nojaf) [inline src/Fantomas.Core/CodePrinter.fs:1859]: Should be a partial active pattern here.

File: src/Fantomas.Core/CodePrinter.fs [hunk 7]
```diff
@@ -1878,9 +1893,8 @@ let genTupleMultiline (node: ExprTupleNode) =
         function
         | Choice1Of2 e ->
             match e with
-            | Expr.IfThen _
-            | Expr.IfThenElif _
-            | Expr.IfThenElse _ when (idx < lastIndex) -> autoParenthesisIfExpressionExceedsPageWidth (genExpr e)
+            | _ when isSynExprIfThenElse e && (idx < lastIndex) ->
```
Comment (nojaf) [inline src/Fantomas.Core/CodePrinter.fs:1896]: Should be a partial active pattern. It is ok to have the when clause for `when (idx < lastIndex)`.

File: src/Fantomas.Core/CodePrinter.fs [hunk 8]
```diff
@@ -2058,9 +2072,10 @@ let genControlExpressionStartCore
 let genMultilineInfixExpr (node: ExprInfixAppNode) =
     let genLhs (ctx: Context) =
         match node.LeftHandSide with
-        | Expr.IfThen _
-        | Expr.IfThenElse _
-        | Expr.IfThenElif _ when (ctx.Config.IndentSize - 1 <= node.Operator.Text.Length) ->
+        | _ when
```
Comment (nojaf) [inline src/Fantomas.Core/CodePrinter.fs:2075]: Should be a partial active pattern. It is ok to have the when clause for `when (ctx.Config.IndentSize - 1 <= node.Operator.Text.Length)`.
Maybe we even want that as a separate let binding to keep the `when` short.

File: src/Fantomas.Core/CodePrinter.fs [hunk 9]
```diff
@@ -1834,15 +1828,36 @@ let genMultilineFunctionApplicationArguments (argExpr: Expr) =
     | _ -> genExpr argExpr
 
 let genTuple (node: ExprTupleNode) =
+    let wrapInfixAppRhsInParenIfNeeded expr =
```
Comment (nojaf) [inline src/Fantomas.Core/CodePrinter.fs:1831]: Maybe add some comments on why we do this.
The `IfNeeded` doesn't really give the why away.

File: src/Fantomas.Core/CodePrinter.fs [hunk 10]
```diff
@@ -1834,15 +1828,36 @@ let genMultilineFunctionApplicationArguments (argExpr: Expr) =
...
+        match expr with
```
Comment (nojaf) [inline src/Fantomas.Core/CodePrinter.fs:1832]: Can you check what happens for `SameInfixApps`?

**Review by nojaf — CHANGES_REQUESTED** [review]
I did some good old fashion nitpicking but know that you are on track here.

File: src/Fantomas.Core/CodePrinter.fs (hunk 10 above)
Comment (dawedawe) [reply src/Fantomas.Core/CodePrinter.fs:1832]: They don't seem to be affected by the issue
See: [here](https://fsprojects.github.io/fantomas-tools/#/fantomas/main?data=N4KABGBEDGD2AmBTSAuKAbRAXMBDMAvGAAwA6AdpjgEaFhYBOArohVWNHQIxvZjx0ATBV44A%2BoQoQIjAJ5TpixizAAfAHxgAlgI3bdmnWsPwANGGoKIAdy1YAFmEQAPMAFp1VxWABmudADOiOZ%2Bgcim4FBBWFha5ADmAahgANqREKDeUAAkWLIADshokHFYkBFZkNkAbv4syWlZEMQVTZAAkuRI5FgAyloAXuHp3gAsIxAAuiMAvq0ZEzl5hcklPeWLVbXo9WiNTVzz3pAAsrjOADJxiBeICQ4bTWBcgsSL04pzI5nHuQVFUDu8AA8j4ruREL08phHr9trtUosIIIjopIABRLqg8HDJ6QdA%2BSDvWZHH5ov4rYrUWCwdCw8nwgH7LIAZlR0g65CCDCwADE4v4AHKIazoa707zKRDEz6kzYUgGQam0iUQLZ1JlIsCjdlq3r5XDQRAAIUQPlgDEQAAVcAxcABbbCIBiq6RSmXSL6KMkchWrZV09nqnaap4AVl1UH1hpNZotN1g1md0FwQU61VgKdisHIrpkzGlWQ%2BnrllT9VJpgfljIaWoAbJHINGjabzZaAKr5QoMFNp8gZrNaHN53z%2BIIeiBe6Q%2BtXlqAB13BhHM7wAdkbzdjbcQAGF0KmAjucwFlNAsBaR6Fx0WSd95ctFQugzUNbWngAODcGltxy0nRD2tQzqXmOhbeMWk6lr8D7%2BpWi4viGb5NAAnF%2BMatvGR7oMOkZXmBigQWAU4LGWMEViqz41nsWpcC0WpNt%2BiAAII%2BFgzpHva9q4CO7o3rKd6kf8sEUdWr7UU8XCHPRm4YZavQAVocDYbmuGgRORFQeSZHznBlFiYiEkotJjEsWxDDyfaim0jhWq8eBt7eveQnkVWZZUQZBxssZMZMQwsBMF0AAiiBipZZk8QW6nEWAM5LM5UClPB7krooXA6vRZzOO0PgACr2HcvT2BaWAAOo6A8kZvHxJYCdB8VrGUemIeJBwRhl5zZXldzomEhXFWV8AVfWVX2fxjmCZSCXrE1y40Q27VZeQPhaM4wLdrg54MOizj5JaAQBEOKlau%2BI0EQ505OZNDVJfpKXSFw64LQASogcAMPAA1DU8oyndIhHRbFVTaddM2hgcn7Pa9FrwIKTCAc6oLtGx9pJJGXBRZpvrA-aTDoLEYoQryFpcTEzo5UJoNIVkXCofRL1vfAJy4-j1xEwwJPhY20D2LahpmWItiDfYRLVZBtVafViWUy1WSvI2mW%2BXasjAgwVwnp9wuRidGPi1jkvTaJzUebLUl4grDBKyratYLD8MMIjyOozROvjXVV043jWgE4gbMc2TFOG7NTyCEZeKK7gyuq1oJ5M573u%2BxtnP0dzvNns6AvlcLLsXRNipS4HYOy15ZvnAAahqxpxPAcTxBrI7a6LGm67OwP525t1aoI6Ul84vIBWeh2V10Nd15GP3ZyRbt5wb7dG3dyJtT3gWwFgADi2DbbtiD7Ydo-Hb9UznZPEtXU%2BBdU94gjzXiwqitcppYEmdzk4UwXLeQdiHUxXT-nbTtPHZM6Y0c5T2Eq5OEHdg6PTDmKeI5A%2B7kAHjmfocCNpMEtDlWAnRuhYA2odECYQJ4xUuo%2BXS58ZaXwhjAsy5A8HVATAkX%2BQEGBvziJ-Y8BDryjRqq7E%2BpCRKzyDk0QQtMzbMy0DiC4DpqDwFwHuWAB0Ei329pw-Cf0j7ENzmAm6c8tQsjoniTezotCOh6P4AA0ogRA%2BRsF3CwJ0Y0dpEGay1HhIhgM5xKjIYIwu3gWSmzaMafc5AADWOIAi%2BX8l0YUJ5ECM3Ed7Tee0DocMjIA9RwDj561Pt4iBuingslDoE20slECBWjtABgJiBRsXgO2D%2BOZgrQH3HabMR0nhuMbgDEhqwPYswhI4w0ITsBQlkDCaWxs-HFzaLHfpJo7TQGGX0aEuI2iVIdIUeAItuFi14dk6ejVyGTMUCybubRLHWMyrbZhoIgm4FCeEkctED5gH%2BpjFu9Uz4%2BIvicxebRlH31-IgWZXtrgcXyEwXBbSknbxSe0po6TD6ZM0aAlyOihGsmvm0IxVTTG4PQD1SyAQXEdLUl0yIkxWglACLyA6yQ3EzCAA)
That's still valid [code](https://sharplab.io/#v2:DYLgZgzgNALiBOBXAdgH2AUxgAgIbYF5sAGAWAChMcAjQ7GJDCq7AYzoEZmtsATOgEwVuOAPqEK2KfXgBPSdMUNEGbKgB82AJb8N23Zp1rDvKNmoKpAdy0wAFtgwAPbAFp1lxdjC5gEDGY+fhhAA).

For the object instantiation case, the compiler forces parens itself [see](https://sharplab.io/#v2:DYLgZgzgNALiBOBXAdgHxgTwA4FMAEAwgBYCGWMO8AFAJR4C8AsAFB5t4C2OHARpXgDcSwPABUAljGA4QeCDHjjkAczwB7cuLXIGeAKrIAxkRyGA1jgAmAOks4wJRMBhqwAHgD6APjwB3SUR4yjgwUHIhLOyc3HzwgsL68KByCkqq9PpGJuZWtvaOzq6ePv4wgcGh4TAsLNIweIa6xGQUcVSR7BJS+EysUexUAMpqXHgARGN4qD7illMzc9N4szQd/XhQa/16SboTNEA)
I wasn't able to come up with a format, in which the parens could be dropped.

File: src/Fantomas.Core/CodePrinter.fs [hunk 11]
```diff
@@ -300,14 +300,27 @@ let genInheritConstructor (ic: InheritConstructor) =
         genType node.Type
         +> sepSpaceOrIndentAndNlnIfExpressionExceedsPageWidth (genExpr node.Expr)
 
-let isSynExprLambdaOrIfThenElse (e: Expr) =
+let genParenExpr openingParen e closingParen r =
```
Comment (nojaf) [inline src/Fantomas.Core/CodePrinter.fs:303]: `gen` is a misleading name here.
You are constructing a new ExprParenNode here, not generating new events inside the `Context`.
Maybe go with some like `mkExprParenNode`. Just avoid `gen` here.

File: src/Fantomas.Core/CodePrinter.fs [hunk 12]
```diff
@@ -300,14 +300,27 @@ let genInheritConstructor (ic: InheritConstructor) =
...
     | Expr.IfThenElif _
     | Expr.IfThenElse _ -> true
     | _ -> false
 
+let (|IsIfThenElse|_|) (e: Expr) = if isIfThenElse e then Some e else None
+
+let isLambdaOrIfThenElse (e: Expr) =
+    match e with
+    | Expr.Lambda _ -> true
+    | IsIfThenElse _ -> true
```
Comment (nojaf) [inline src/Fantomas.Core/CodePrinter.fs:318]: Nit, combine these in an Or pattern:
```fsharp
    | Expr.Lambda _
    | IsIfThenElse _ -> true
```

**Review by nojaf — APPROVED** [review]
Looks good!

**Review by nojaf — APPROVED** [review]

**Pull request merged**